File: scripts/updateTerritories.js
/**
 * Update territories.js with generated geographic polygon data
 * Run with: node scripts/updateTerritories.js
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load generated polygons
const polygonsPath = path.join(__dirname, 'generated-polygons.json');
const polygons = JSON.parse(fs.readFileSync(polygonsPath, 'utf8'));

// Read territories.js
const territoriesPath = path.join(__dirname, '../src/data/territories.js');
let territoriesContent = fs.readFileSync(territoriesPath, 'utf8');

console.log('Updating territories.js with geographic polygon data...\n');

let updateCount = 0;
let notFoundCount = 0;

// For each generated polygon, update the corresponding territory
Object.entries(polygons).forEach(([territoryId, polygonData]) => {
  // Check if this territory exists in territories.js
  const territoryPattern = new RegExp(`${territoryId}:\\s*{[^}]*id:\\s*['"]${territoryId}['"]`, 'i');

  if (!territoryPattern.test(territoriesContent)) {
    console.log(`⚠ Territory not found in territories.js: ${territoryId}`);
    notFoundCount++;
    return;
  }

  // Find and replace the polygon, centroid, labelPosition, troopPosition, and pointsPosition
  // This regex finds the entire territory object
  const fullTerritoryRegex = new RegExp(
    `(${territoryId}:\\s*{[\\s\\S]*?)` + // Start of territory object
    `(polygon:\\s*{[^}]*},?\\s*)` +       // polygon object
    `([\\s\\S]*?` +                       // content between polygon and centroid
    `centroid:\\s*{[^}]*},?\\s*)` +       // centroid object
    `([\\s\\S]*?` +                       // content between centroid and labelPosition
    `labelPosition:\\s*{[^}]*},?\\s*)` +  // labelPosition object
    `([\\s\\S]*?` +                       // content between labelPosition and troopPosition
    `troopPosition:\\s*{[^}]*},?\\s*)` +  // troopPosition object
    `([\\s\\S]*?` +                       // content between troopPosition and pointsPosition
    `pointsPosition:\\s*{[^}]*},?\\s*)`,  // pointsPosition object
    'i'
  );

  const newPolygon = `polygon: {\n      d: '${polygonData.d}',\n    },`;
  const newCentroid = `centroid: { x: ${polygonData.centroid.x.toFixed(2)}, y: ${polygonData.centroid.y.toFixed(2)} },`;
  const newLabelPos = `labelPosition: { x: ${polygonData.labelPosition.x.toFixed(2)}, y: ${polygonData.labelPosition.y.toFixed(2)} },`;
  const newTroopPos = `troopPosition: { x: ${polygonData.troopPosition.x.toFixed(2)}, y: ${polygonData.troopPosition.y.toFixed(2)} },`;
  const newPointsPos = `pointsPosition: { x: ${polygonData.pointsPosition.x.toFixed(2)}, y: ${polygonData.pointsPosition.y.toFixed(2)} },`;

  territoriesContent = territoriesContent.replace(
    fullTerritoryRegex,
    `$1${newPolygon}\n    $3${newCentroid}\n    $5${newLabelPos}\n    $7${newTroopPos}\n    $9${newPointsPos}\n    `
  );

  updateCount++;
  console.log(`✓ Updated ${territoryId}`);
});

// Write back to territories.js
fs.writeFileSync(territoriesPath, territoriesContent);

console.log(`\n✓ Updated ${updateCount} territories`);
if (notFoundCount > 0) {
  console.log(`⚠ ${notFoundCount} territories not found in territories.js`);
}
console.log('\nterritories.js has been updated with geographic polygon data');

File: scripts/update_territories.py
#!/usr/bin/env python3
"""
Update territories.js with generated geographic polygon data
"""

import json
import re

# Load generated polygons
with open('scripts/generated-polygons.json', 'r') as f:
    polygons = json.load(f)

# Read territories.js
with open('src/data/territories.js', 'r') as f:
    content = f.read()

print('Updating territories.js with geographic polygon data...\n')

# For each territory, update its data
for territory_id, data in polygons.items():
    print(f'Updating {territory_id}...')

    # Create regex pattern to match the entire territory object
    # Pattern: "  territoryId: {" ... "  },"
    pattern = rf'(  {territory_id}:\s*{{.*?)(  }},)'

    match = re.search(pattern, content, re.DOTALL)
    if not match:
        print(f'  ⚠ Could not find {territory_id}')
        continue

    territory_block = match.group(1)

    # Replace polygon.d
    territory_block = re.sub(
        r"(polygon:\s*{\s*d:\s*)'[^']*'",
        rf"\1'{data['d']}'",
        territory_block
    )

    # Replace centroid
    territory_block = re.sub(
        r'centroid:\s*{[^}]*}',
        f"centroid: {{ x: {data['centroid']['x']:.2f}, y: {data['centroid']['y']:.2f} }}",
        territory_block
    )

    # Replace labelPosition
    territory_block = re.sub(
        r'labelPosition:\s*{[^}]*}',
        f"labelPosition: {{ x: {data['labelPosition']['x']:.2f}, y: {data['labelPosition']['y']:.2f} }}",
        territory_block
    )

    # Replace troopPosition
    territory_block = re.sub(
        r'troopPosition:\s*{[^}]*}',
        f"troopPosition: {{ x: {data['troopPosition']['x']:.2f}, y: {data['troopPosition']['y']:.2f} }}",
        territory_block
    )

    # Replace pointsPosition
    territory_block = re.sub(
        r'pointsPosition:\s*{[^}]*}',
        f"pointsPosition: {{ x: {data['pointsPosition']['x']:.2f}, y: {data['pointsPosition']['y']:.2f} }}",
        territory_block
    )

    # Replace in content
    content = re.sub(pattern, territory_block + '  },', content, count=1, flags=re.DOTALL)

    print(f'  ✓ Updated')

# Write back
with open('src/data/territories.js', 'w') as f:
    f.write(content)

print('\n✓ Updated territories.js with geographic polygon data')
